### PR TITLE
Reject requests if corruption

### PR DESCRIPTION
**Pull request by sudiksha27**
**Goals (and why)**:
Final piece of TimeLock shooting itself upon detecting local/ remote corruption.

**Implementation Description (bullets)**:

- Added a Jersey filter + Undertow handler to reject requests at RPC level if detected corruption.

- Wrap undertow endpoints in corruption handler while registering undertow resources.

**Testing (What was existing testing like?  What have you done to improve it?)**:

- Tested on internal test stack

**Concerns (what feedback would you like?)**:

- Have I missed out something in filter + handler implementation?

**Where should we start reviewing?**:
UndertowCorruptionHandlerService.java, JerseyCorruptionFilter.java

**Priority (whenever / two weeks / yesterday)**:
EOD Wednesday would be great.

### CODE REVIEW

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/UndertowCorruptionHandlerService.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    public UndertowCorruptionHandlerService(UndertowService service, CorruptionHealthCheck healthCheck) {
+        this.delegate = service;
+        this.healthCheck = healthCheck;
+        this.wrapper = handler -> exchange -> {
+            if (healthCheck.isHealthy()) {
+                handler.handleRequest(exchange);
+            } else {
+                throw new ServiceNotAvailableException("TimeLock is not available on account of data corruption.");
+            }
```
Comment (sudiksha27) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/UndertowCorruptionHandlerService.java:42]: remote exceptions - see what alta does

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/UndertowCorruptionHandlerService.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    private final CorruptionHealthCheck healthCheck;
+
+    public UndertowCorruptionHandlerService(UndertowService service, CorruptionHealthCheck healthCheck) {
+        this.delegate = service;
+        this.healthCheck = healthCheck;
+        this.wrapper = handler -> exchange -> {
+            if (healthCheck.isHealthy()) {
+                handler.handleRequest(exchange);
+            } else {
+                throw new ServiceNotAvailableException("TimeLock is not available on account of data corruption.");
```
Comment (sudiksha27) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/UndertowCorruptionHandlerService.java:41]: Note - QosException.unavailable returns 200 response and the client may retry at a later time.

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/Constants.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+package com.palantir.atlasdb.timelock.corruption;
+
+public final class Constants {
+    private Constants() {
+        // no op
+    }
+
+    public static final String CORRUPTION_MESSAGE
+            = "{\"error\": \"TimeLock is not available on account of data corruption.\"}";
```
Comment (jeremyk-91) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/Constants.java:25]: This should be consistent with the other QoS exceptions we get.

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/UndertowCorruptionHandlerService.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    public UndertowCorruptionHandlerService(UndertowService service, CorruptionHealthCheck healthCheck) {
+        this.delegate = service;
+        this.healthCheck = healthCheck;
+        this.wrapper = handler -> exchange -> {
+            if (healthCheck.isHealthy()) {
+                handler.handleRequest(exchange);
+            } else {
+                exchange.setStatusCode(StatusCodes.SERVICE_UNAVAILABLE);
+                exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "application/json");
```
Comment (jeremyk-91) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/UndertowCorruptionHandlerService.java:43]: I don't think this is needed if we standardise with Witchcraft.

**Review by jeremyk-91 — APPROVED** [review]
👍

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/UndertowCorruptionHandlerService.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    private final CorruptionHealthCheck healthCheck;
+
+    public UndertowCorruptionHandlerService(UndertowService service, CorruptionHealthCheck healthCheck) {
+        this.delegate = service;
+        this.healthCheck = healthCheck;
+        this.wrapper = handler -> exchange -> {
+            if (healthCheck.isHealthy()) {
+                handler.handleRequest(exchange);
+            } else {
+                exchange.setStatusCode(StatusCodes.SERVICE_UNAVAILABLE);
```
Comment (j-baker) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/corruption/UndertowCorruptionHandlerService.java:42]: as a note, you can here `throw QosException.unavailabe()` and induce the same behaviour

**Pull request merged**